File: day06/main.ml
let read_lines path =
    let ic = open_in path in
    let rec read_all_lines () =
        try
            let line = input_line ic in
            line :: read_all_lines ()
        with End_of_file ->
            close_in ic;
            []
    in
    read_all_lines ()

type action = On | Off | Toggle

type pos = {
    x : int;
    y : int;
}

let parse_pos (input : string) : pos =
    Scanf.sscanf input "%d,%d" (fun x y -> { x; y })

type range = {
    src : pos;
    dst : pos;
}

type inst = {
    action : action;
    range : range;
}

let parse_inst (input : string) : inst =
    match String.split_on_char ' ' input with
    | "toggle"        :: src :: "through" :: dst :: [] -> { action = Toggle; range = { src = parse_pos src; dst = parse_pos dst } }
    | "turn" :: "on"  :: src :: "through" :: dst :: [] -> { action = On;     range = { src = parse_pos src; dst = parse_pos dst } }
    | "turn" :: "off" :: src :: "through" :: dst :: [] -> { action = Off;    range = { src = parse_pos src; dst = parse_pos dst } }
    | _ -> raise (Invalid_argument input)

let in_range ({ x : int; y : int } : pos) ({ src : pos; dst : pos } : range) : bool =
    let lx = min src.x dst.x in
    let ly = min src.y dst.y in
    let hx = max src.x dst.x in
    let hy = max src.y dst.y in
    lx <= x && x <= hx && ly <= y && y <= hy

let part1 (insts : inst list) : int =
    let rec find_state (pos : pos) (insts : inst list) : bool =
        match insts with
        | [] -> false
        | { action; range } :: rest ->
            if not (in_range pos range) then
                find_state pos rest
            else
                if action == Toggle then
                    not (find_state pos rest)
                else
                    action == On
    in

    let insts = List.rev insts in

    let count = ref 0 in
    for x = 0 to 999 do
        for y = 0 to 999 do
            if find_state { x; y } insts then
                incr count
        done
    done;
    !count

let part2 (insts : inst list) : int =
    let brightness = function
        | On -> 1
        | Off -> -1
        | Toggle -> 2
    in

    let rec find_bright (pos : pos) (insts : inst list) (current : int) : int =
        match insts with
        | [] -> current
        | { action; range } :: rest ->
            if not (in_range pos range) then
                find_bright pos rest current
            else
                let bright = brightness action in
                max 0 (current + bright) |> find_bright pos rest
    in

    let total = ref 0 in
    for x = 0 to 999 do
        for y = 0 to 999 do
            let bright = find_bright { x; y } insts 0 in
            total := !total + bright
        done
    done;
    !total


let () =
    let insts = read_lines "./input.txt" |> List.map parse_inst in
    Printf.printf "part 1: %d\n" (part1 insts);
    flush stdout;
    Printf.printf "part 2: %d\n" (part2 insts);

File: day06/Main.hs
{-# LANGUAGE DerivingStrategies #-}

import Data.Bits (xor)
import Data.Bool (bool)

main :: IO ()
main = do
  insts <- map parseInst . lines <$> readFile "./input.txt"
  putStrLn $ unwords ["part 1:", show (part1 insts)]
  putStrLn $ unwords ["part 2:", show (part2 insts)]

part1 :: [Inst] -> Int
part1 insts = length [() | x <- [0..999], y <- [0..999], findState (Pos x y) (reverse insts)]

part2 :: [Inst] -> Int
part2 insts = getBrightness $ sum $ map (`findBrightness` insts) $ Pos <$> [0..999] <*> [0..999]

findState :: Pos -> [Inst] -> Bool
findState pos insts =
  case span (==Toggle) actions of
    (toggles, state:_) -> xor (state == On) (odd $ length toggles)
    (toggles, _)       -> odd $ length toggles
  where
    actions = map instAction $ filter (inRange pos . instRange) insts

findBrightness :: Pos -> [Inst] -> Brightness
findBrightness pos = sum . map (convertBright . instAction) . filter (inRange pos . instRange)

newtype Brightness = Brightness { getBrightness :: Int }
  deriving newtype (Show, Eq, Ord)

instance Num Brightness where
  Brightness a + Brightness b = Brightness (max (a + b) 0)
  (-) = undefined
  (*) = undefined
  abs = undefined
  signum = undefined
  fromInteger = Brightness . fromInteger

convertBright :: Action -> Brightness
convertBright Toggle = Brightness 2
convertBright On = Brightness 1
convertBright Off = Brightness (-1)

data Inst = Inst
  { instAction :: Action
  , instRange  :: Range
  } deriving Show

parseInst :: String -> Inst
parseInst s =
  case words s of
    ["toggle", from, "through", to]      -> Inst Toggle $ Range (parsePos from) (parsePos to)
    ["turn", "on", from, "through", to]  -> Inst On $ Range (parsePos from) (parsePos to)
    ["turn", "off", from, "through", to] -> Inst Off $ Range (parsePos from) (parsePos to)

data Action = On | Off | Toggle
  deriving (Show, Eq)

data Range = Range
  { rangeFrom :: Pos
  , rangeTo   :: Pos
  } deriving Show

-- inclusive
inRange :: Pos -> Range -> Bool
inRange (Pos x y) (Range (Pos x1 y1) (Pos x2 y2)) = lx <= x && x <= hx && ly <= y && y <= hy
  where
    lx = min x1 x2
    ly = min y1 y2
    hx = max x1 x2
    hy = max y1 y2

data Pos = Pos
  { posX :: Int
  , posY :: Int
  } deriving Show

parsePos :: String -> Pos
parsePos s = Pos (read x_str) (read y_str)
  where
    (x_str, ',':y_str) = span (/=',') s
